File: go/allergies/allergies.go
package allergies

const testVersion = 1

type allergyType uint

const (
	eggs allergyType = 1 << iota
	peanuts
	shellfish
	strawberries
	tomatoes
	chocolate
	pollen
	cats
)

var allergyTypeToString = map[allergyType]string{
	eggs:         "eggs",
	peanuts:      "peanuts",
	shellfish:    "shellfish",
	strawberries: "strawberries",
	tomatoes:     "tomatoes",
	chocolate:    "chocolate",
	pollen:       "pollen",
	cats:         "cats"}

func Allergies(input uint) (allergies []string) {
	for allergy := range allergyTypeToString {
		if input&uint(allergy) != 0 {
			allergies = append(allergies, allergyTypeToString[allergy])
		}
	}
	return allergies
}

func AllergicTo(input uint, allergen string) bool {
	allergies := Allergies(input)
	for _, a := range allergies {
		if a == allergen {
			return true
		}
	}
	return false
}

File: haskell/allergies/src/Allergies.hs
module Allergies
  ( Allergen(..)
  , allergies
  , isAllergicTo
  ) where

import           Data.Bits

data Allergen
  = Eggs
  | Peanuts
  | Shellfish
  | Strawberries
  | Tomatoes
  | Chocolate
  | Pollen
  | Cats
  deriving (Eq, Enum)

allergies :: Int -> [Allergen]
allergies score
  | popCount (score .&. bit 0) == 1 = Eggs : allergies (score - bit 0)
  | popCount (score .&. bit 1) == 1 = Peanuts : allergies (score - bit 1)
  | popCount (score .&. bit 2) == 1 = Shellfish : allergies (score - bit 2)
  | popCount (score .&. bit 3) == 1 = Strawberries : allergies (score - bit 3)
  | popCount (score .&. bit 4) == 1 = Tomatoes : allergies (score - bit 4)
  | popCount (score .&. bit 5) == 1 = Chocolate : allergies (score - bit 5)
  | popCount (score .&. bit 6) == 1 = Pollen : allergies (score - bit 6)
  | popCount (score .&. bit 7) == 1 = Cats : allergies (score - bit 7)
  | otherwise = []

isAllergicTo :: Allergen -> Int -> Bool
isAllergicTo allergen score = allergen `elem` allergies score
